src/libc/stdio/fgetws.3
.\" $NetBSD: fgetws.3,v 1.1 2003/03/07 07:11:36 tshiozak Exp $
.\"
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek and the American National Standards Committee X3,
.\" on Information Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)fgets.3	8.1 (Berkeley) 6/4/93
.\"
.\" Original version ID:
.\" FreeBSD: src/lib/libc/stdio/fgets.3,v 1.16 2002/05/31 05:01:17 archie Exp
.\" FreeBSD: src/lib/libc/stdio/fgetws.3,v 1.2 2002/09/06 11:23:55 tjr Exp
.\"
.Dd August 6, 2002
.Dt FGETWS 3
.Os
.Sh NAME
.Nm fgetws
.Nd get a line of wide characters from a stream
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdio.h
.In wchar.h
.Ft "wchar_t *"
.Fn fgetws "wchar_t * restrict ws" "int n" "FILE * restrict fp"
.Sh DESCRIPTION
The
.Fn fgetws
function
reads at most one less than the number of characters specified by
.Fa n
from the given
.Fa fp
and stores them in the wide character string
.Fa ws .
Reading stops when a newline character is found,
at end-of-file or error.
The newline, if any, is retained.
If any characters are read and there is no error, a
.Ql \e0
character is appended to end the string.
.Sh RETURN VALUES
Upon successful completion,
.Fn fgetws
returns
.Fa ws .
If end-of-file occurs before any characters are read,
.Fn fgetws
returns
.Dv NULL
and the buffer contents remain unchanged.
If an error occurs,
.Fn fgetws
returns
.Dv NULL
and the buffer contents are indeterminate.
The
.Fn fgetws
function
does not distinguish between end-of-file and error, and callers must use
.Xr feof 3
and
.Xr ferror 3
to determine which occurred.
.Sh ERRORS
.Bl -tag -width Er
.It Bq Er EBADF
The given
.Fa fp
argument is not a readable stream.
.It Bq Er EILSEQ
The data obtained from the input stream does not form a valid
multibyte character.
.El
.Pp
The function
.Fn fgetws
may also fail and set
.Va errno
for any of the errors specified for the routines
.Xr fflush 3 ,
.Xr fstat 2 ,
.Xr read 2 ,
or
.Xr malloc 3 .
.Sh SEE ALSO
.Xr feof 3 ,
.Xr ferror 3 ,
.Xr fgets 3
.Sh STANDARDS
The
.Fn fgetws
function
conforms to
.St -p1003.1-2001 .


RELEASE-NOTES
Release notes for BitKeeper version 4.1 (released 12-Oct-2007)

Major features
    BAM support.  BAM stands for "Binary Asset Management" and it adds 
    support to BK for versioning large binaries.  It solves two problems:
    a) one or more binary files that are frequently changed.
    b) collections of many large binaries where you only need a subset.
    The way it solves this is to introduce the concept of BAM server[s].
    A BAM server manages a collection of binaries for one or more BAM
    clients.  BAM clients may have no data present; when it is needed
    the data is fetched from the BAM server.  
    
    In the first case above, only the tip will be fetched.  Imagine that
    you have 100 deltas, each 10MB in size.  The history is 1GB but you
    only need 10MB in your clone.

    In the second case, imagine that you have thousands of game assets
    distributed across multiple directories.  You typically work only
    in one directory at a time.  You will only need to fetch the subset
    of files that you need, the rest of the repository will have the
    history of what changed but no data (so bk log will work but 
    bk cat will have to go fetch the data).

    BAM files are currently limited to 32 bits in size.

    See bk help Howto-BAM and bk help bam for more information.

    -----

    bk repair makes use of the new -@<url> and BAM infrastructure to run
    much more quickly.  Our stress tests for this amount to a loop, cloning
    each tagged version of BitKeeper, removing each top level directory,
    and running bk repair to restore.  Works fine, and works quickly.
    Please let us know if you find problems with this feature, for now it
    is enabled when you run check with -ff (2 "f" chars) but in the future
    we intend to turn it on in any repo with autofix configured.
    See bk help repair for more information.

Summary of changes

Performance fixes
    Repositories with large numbers of changesets will run faster when
    doing pull/push/changes.  For example, in the Linux kernel 2.6 tree
    with 128 thousand changesets, the time to do "bk changes -r+" is 1/2
    of what it was in bk-4.0.2.  Null pulls/pushes should see similar
    performance gains.

    Pulls which update repositories with large numbers of deleted files
    will run significantly faster.

Windows Vista support
    The Windows version of BitKeeper now runs on Windows Vista on 32-bit
    architectures.  64-bit support for Vista and/or XP will be in a future
    release.

Triggers
    This release introduces the concept of a trigger path.  The trigger
    path may be set in the configuration file to create a set of directories
    to search.  All triggers found in all directories are run (unless one
    is a pre- trigger and it fails).
    See bk help config-etc.

    This release introduces support for pre-undo triggers.  
    See bk help triggers.

Checkout mode
    A new checkout mode, checkout:last, has been added.  For clones, it
    is the same as checkout:none, but after the file has been checked out
    (read-only or read-write), BK will maintain that state.  The main
    use is for repositories with large numbers of large BAM binaries.
    The mode makes it possible to work on a subset of those files which
    gives you the effect of a sparse tree.  If said repository was in
    checkout:get or checkout:edit mode then the tips of all files will
    be present, i.e., the repository will not be sparse.

BK/Web
    BK/Web now uses MD5 keys for all URLs and revision numbers.  This
    allows bookmarks to continue to work even after revision numbers are
    changed by merges.  For this reason, the "Link to this page" links have
    been removed.
    Note the URL syntax has not changed from the 4.0 release, so old URLs
    will still work correctly.
    
    BK/Web now lists extra files while browsing the source tree; they are
    at the end of the directory listing which is { dirs, BK files, extras }.
    Extra files are only listed if they are not ignored.  Extra files 
    may be clicked on and viewed.

    BK/Web has links on the file names in the history/annotate pages that
    will let you download a copy of that file with no markups.

SSH URLs
    BitKeeper now allows an SSH URL to specify a non-standard port,
    for example ssh://host.example.com:5555/repopath.  See bk help url.

Remote command execution
    The command syntax 'bk -@URL command' allows a bk command to be
    run on a remote repository with the output returned locally.
    See 'bk help bk' for more information.
    Example showing work in progress (diffs and new files) in a remote repo:

    	bk -@bk://work/BAM -r diffs -Nur

Configuration
    nosync: on

    By default, BK does a sync in the resolver after updating your repository.
    The reason it does not fsync() individual files is that on some Linux
    file systems, the fsync() is like a sync() and sync() is global.  Doing
    it over and over again hurt server performance.  So we do one sync after
    applying all the files.

    Some customers prefer performance over safety; they can set this variable
    and get no syncs at all.  We don't recommend that.

bk admin
    The -H option formerly checked sfile structure, limited to AT&T
    SCCS features, and checked whether the file contents were ASCII.
    As BitKeeper now fully supports non-ASCII characters, this option
    is deprecated.  Use -h, -hh, or -hhh instead.  See bk help admin.

bk bkd
    The -p option now allows you to restrict allowed connections to
    those originating from a specified IP address.  A common usage is
    "-p127.0.0.1:", which accepts connections only on the default port
    (14690) on localhost.  See bk help bkd.

bk changes
    The search facility ("-/<pattern>/") now supports two additional
    suffixes:  "g", which indicates that <pattern> is a glob, not a regular
    expression; and "t", which restricts the search to tag names.  See bk
    help changes.


src/libc/stdio/fgetws.c
/* $NetBSD: fgetws.c,v 1.1 2003/03/07 07:11:37 tshiozak Exp $ */

/*-
 * Copyright (c) 2002 Tim J. Robbins.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * Original version ID:
 * FreeBSD: src/lib/libc/stdio/fgetws.c,v 1.4 2002/09/20 13:25:40 tjr Exp
 *
 */

#if defined(LIB_SCCS) && !defined(lint)
__RCSID("$NetBSD: fgetws.c,v 1.1 2003/03/07 07:11:37 tshiozak Exp $");
#endif

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <wchar.h>
#include "reentrant.h"
#include "local.h"

wchar_t *
fgetws(ws, n, fp)
	wchar_t * __restrict ws;
	int n;
	FILE * __restrict fp;
{
	wchar_t *wsp;
	wint_t wc;

	assert(fp != NULL);
	assert(ws != NULL);

	FLOCKFILE(fp);

	if (n <= 0) {
		errno = EINVAL;
		goto error;
	}

	wsp = ws;
	while (n-- > 1) {
		wc = __fgetwc_unlock(fp);
		if (__sferror(fp) != 0)
			goto error;
		if (__sfeof(fp) != 0) {
			if (wsp == ws) {
				/* EOF/error, no characters read yet. */
				goto error;
			}
			break;
		}
		*wsp++ = (wchar_t)wc;
		if (wc == L'\n') {
			break;
		}
	}

	*wsp++ = L'\0';
	FUNLOCKFILE(fp);

	return (ws);

error:
	FUNLOCKFILE(fp);
	return (NULL);
}
